File: src/go/nyx/test/integration/services/service_factory_test.go
//go:build integration
// +build integration

// Only run these tests as part of the integration test suite, when the 'integration' build flag is passed (i.e. running go test --tags=integration)

/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package services_test

import (
	"testing" // https://pkg.go.dev/testing

	log "github.com/sirupsen/logrus"            // https://pkg.go.dev/github.com/sirupsen/logrus
	assert "github.com/stretchr/testify/assert" // https://pkg.go.dev/github.com/stretchr/testify/assert

	ent "github.com/mooltiverse/nyx/src/go/nyx/entities"
	svc "github.com/mooltiverse/nyx/src/go/nyx/services"
	svcapi "github.com/mooltiverse/nyx/src/go/nyx/services/api"
)

var (
	// use this slice to parametrize tests based on the providers
	serviceProviders = []ent.Provider{
		ent.GITHUB,
		ent.GITLAB,
	}

	// use this slice to parametrize tests based on the features
	serviceFeatures = []svcapi.Feature{
		svcapi.GIT_HOSTING,
		svcapi.RELEASES,
		svcapi.RELEASE_ASSETS,
		svcapi.USERS,
	}
)

func TestServiceFactoryGitHostingServiceInstance(t *testing.T) {
	logLevel := log.GetLevel()   // save the previous logging level
	log.SetLevel(log.ErrorLevel) // set the logging level to filter out warnings produced during tests

	for _, p := range serviceProviders {
		t.Run(p.String(), func(t *testing.T) {
			_, err := svc.GitHostingServiceInstance(p, map[string]string{})
			assert.NoError(t, err)
		})
	}

	log.SetLevel(logLevel) // restore the original logging level
}

func TestServiceFactoryInstance(t *testing.T) {
	logLevel := log.GetLevel()   // save the previous logging level
	log.SetLevel(log.ErrorLevel) // set the logging level to filter out warnings produced during tests

	for _, p := range serviceProviders {
		t.Run(p.String(), func(t *testing.T) {
			_, err := svc.Instance(p, map[string]string{})
			assert.NoError(t, err)
		})
	}

	log.SetLevel(logLevel) // restore the original logging level
}

func TestServiceFactoryReleaseServiceInstance(t *testing.T) {
	logLevel := log.GetLevel()   // save the previous logging level
	log.SetLevel(log.ErrorLevel) // set the logging level to filter out warnings produced during tests

	for _, p := range serviceProviders {
		t.Run(p.String(), func(t *testing.T) {
			_, err := svc.ReleaseServiceInstance(p, map[string]string{})
			assert.NoError(t, err)
		})
	}

	log.SetLevel(logLevel) // restore the original logging level
}

func TestServiceFactoryUserServiceInstance(t *testing.T) {
	logLevel := log.GetLevel()   // save the previous logging level
	log.SetLevel(log.ErrorLevel) // set the logging level to filter out warnings produced during tests

	for _, p := range serviceProviders {
		t.Run(p.String(), func(t *testing.T) {
			_, err := svc.UserServiceInstance(p, map[string]string{})
			assert.NoError(t, err)
		})
	}

	log.SetLevel(logLevel) // restore the original logging level
}

File: src/java/main/src/integrationTest/java/com/mooltiverse/oss/nyx/services/ServiceFactoryTest.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.services;

import static org.junit.jupiter.api.Assertions.*;

import java.util.Map;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.junit.jupiter.params.provider.NullSource;

import com.mooltiverse.oss.nyx.entities.Provider;
import com.mooltiverse.oss.nyx.services.github.GitHub;
import com.mooltiverse.oss.nyx.services.gitlab.GitLab;

@DisplayName("ServiceFactory")
public class ServiceFactoryTest {
    @Nested
    @DisplayName("ServiceFactory.gitHostingServiceInstance()")
    class GitHostingServiceInstanceTests {
        @ParameterizedTest(name = "ServiceFactory.gitHostingServiceInstance(''{0}'', Map<String,Object>).getClass() == ''{0}''")
        @EnumSource(Provider.class)
        void gitHostingServiceInstance(Provider provider)
            throws Exception {
            if (provider.equals(Provider.GITHUB))
                assertEquals(GitHub.class, ServiceFactory.gitHostingServiceInstance(provider, Map.<String,String>of()).getClass());
            else if (provider.equals(Provider.GITLAB))
                assertEquals(GitLab.class, ServiceFactory.gitHostingServiceInstance(provider, Map.<String,String>of()).getClass());
            else assertThrows(UnsupportedOperationException.class, () -> ServiceFactory.gitHostingServiceInstance(provider, Map.<String,String>of()));
        }

        @ParameterizedTest(name = "ServiceFactory.gitHostingServiceInstance(null, Map<String,Object>).getClass() throws NullPointerException")
        @NullSource()
        void exceptionUsingGitHostingServiceInstanceWithNullProvider(Provider provider)
            throws Exception {
            assertThrows(NullPointerException.class, () -> ServiceFactory.gitHostingServiceInstance(provider, Map.<String,String>of()));
        }
    }

    @Nested
    @DisplayName("ServiceFactory.instance()")
    class InstanceTests {
        @ParameterizedTest(name = "ServiceFactory.instance(''{0}'', Map<String,Object>).getClass() == ''{0}''")
        @EnumSource(Provider.class)
        void instance(Provider provider)
            throws Exception {
            if (provider.equals(Provider.GITHUB))
                assertEquals(GitHub.class, ServiceFactory.instance(provider, Map.<String,String>of()).getClass());
            else if (provider.equals(Provider.GITLAB))
                assertEquals(GitLab.class, ServiceFactory.instance(provider, Map.<String,String>of()).getClass());
            else assertThrows(UnsupportedOperationException.class, () -> ServiceFactory.instance(provider, Map.<String,String>of()));
        }

        @ParameterizedTest(name = "ServiceFactory.instance(null, Map<String,Object>).getClass() throws NullPointerException")
        @NullSource()
        void exceptionUsingInstanceWithNullProvider(Provider provider)
            throws Exception {
            assertThrows(NullPointerException.class, () -> ServiceFactory.instance(provider, Map.<String,String>of()));
        }
    }

    @Nested
    @DisplayName("ServiceFactory.releaseServiceInstance()")
    class ReleaseServiceInstanceTests {
        @ParameterizedTest(name = "ServiceFactory.releaseServiceInstance(''{0}'', Map<String,Object>).getClass() == ''{0}''")
        @EnumSource(Provider.class)
        void releaseServiceInstance(Provider provider)
            throws Exception {
            if (provider.equals(Provider.GITHUB))
                assertEquals(GitHub.class, ServiceFactory.releaseServiceInstance(provider, Map.<String,String>of()).getClass());
            else if (provider.equals(Provider.GITLAB))
                assertEquals(GitLab.class, ServiceFactory.releaseServiceInstance(provider, Map.<String,String>of()).getClass());
            else assertThrows(UnsupportedOperationException.class, () -> ServiceFactory.releaseServiceInstance(provider, Map.<String,String>of()));
        }

        @ParameterizedTest(name = "ServiceFactory.releaseServiceInstance(null, Map<String,Object>).getClass() throws NullPointerException")
        @NullSource()
        void exceptionUsingReleaseServiceInstanceWithNullProvider(Provider provider)
            throws Exception {
            assertThrows(NullPointerException.class, () -> ServiceFactory.releaseServiceInstance(provider, Map.<String,String>of()));
        }
    }

    @Nested
    @DisplayName("ServiceFactory.userServiceInstance()")
    class UserServiceInstanceTests {
        @ParameterizedTest(name = "ServiceFactory.userServiceInstance(''{0}'', Map<String,Object>).getClass() == ''{0}''")
        @EnumSource(Provider.class)
        void userServiceInstance(Provider provider)
            throws Exception {
            if (provider.equals(Provider.GITHUB))
                assertEquals(GitHub.class, ServiceFactory.userServiceInstance(provider, Map.<String,String>of()).getClass());
            else if (provider.equals(Provider.GITLAB))
                assertEquals(GitLab.class, ServiceFactory.userServiceInstance(provider, Map.<String,String>of()).getClass());
            else assertThrows(UnsupportedOperationException.class, () -> ServiceFactory.userServiceInstance(provider, Map.<String,String>of()));
        }

        @ParameterizedTest(name = "ServiceFactory.userServiceInstance(null, Map<String,Object>).getClass() throws NullPointerException")
        @NullSource()
        void exceptionUsingUserServiceInstanceWithNullProvider(Provider provider)
            throws Exception {
            assertThrows(NullPointerException.class, () -> ServiceFactory.userServiceInstance(provider, Map.<String,String>of()));
        }
    }
}